realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/RealmFunctionsTests.kt

```diff
@@ -116,36 +162,205 @@ class RealmFunctionsTests {
-                CodecRegistries.fromProviders(
-                        PojoCodecProvider.builder()
-                                .register(Dog::class.java)
-                                .build()
-                )
-        )
+    fun asyncCallFunction() = looperThread.runBlocking {
+        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(32), Integer::class.java) { result ->
+            if (result.isSuccess) {
+                assertEquals(32, result.get().toInt())
+            } else  {
+                fail()
+            }
+            looperThread.testComplete()
+        }
+    }
+
+
+    @Test
+    fun codecArgumentFailure() {
+        val input = Dog("PojoFido")
+        assertFailsWith<CodecConfigurationException> {
+            functions.callFunction(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java)
+        }
+    }
+
+    @Test
+    fun asyncCodecArgumentFailure() = looperThread.runBlocking {
+        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(Dog("PojoFido")), Integer::class.java) { result ->
+            if (result.isSuccess) {
+                fail()
+            } else  {
+                assertTrue(result.error.exception is CodecConfigurationException)
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun codecResponseFailure() {
+        assertFailsWith<CodecConfigurationException> {
+            functions.callFunction(FIRST_ARG_FUNCTION, listOf(32), Dog::class.java)
+        }
+    }
+
+    @Test
+    fun asyncCodecResponseFailure() = looperThread.runBlocking {
+        val input = Dog("PojoFido")
+        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(Dog("PojoFido")), Integer::class.java) { result ->
+            if (result.isSuccess) {
+                fail()
+            } else  {
+                assertTrue(result.error.exception is CodecConfigurationException)
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun localCodecRegistry() {
+        val input = Dog("PojoFido")
+        assertEquals(input, functions.callFunction(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java, pojoRegistry))
+    }
 
+    @Test
+    fun asyncLocalCodecRegistry() = looperThread.runBlocking {
+        val input = Dog("PojoFido")
+        functions.callFunctionAsync(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java, pojoRegistry) { result ->
+            if (result.isSuccess) {
+                assertEquals(input, result.get())
+            } else  {
+                fail()
+            }
+            looperThread.testComplete()
+        }
+    }
+
+    @Test
+    fun instanceCodecRegistry() {
+        val input = Dog("PojoFido")
+        val functionsWithCodecRegistry = anonUser.getFunctions(pojoRegistry)
+        assertEquals(input, functionsWithCodecRegistry.callFunction(FIRST_ARG_FUNCTION, listOf(input), Dog::class.java))
+    }
+
+    @Test
+    fun unknownFunction() {
+        assertFailsWith<ObjectServerError> {
```

Comment:
We have a better `KotlinTestUtils.expectErrorCode` for this. The name might need a revisit after moving to `assertFailsWith` though.